packages/validator/test/capability-access.spec.js

```diff
@@ -0,0 +1,556 @@
+      audience: w3,
+      capabilities: [
+        {
+          with: alice.did(),
+          can: 'store/add',
+          nb: {
+            link: Link.parse('bafkqaaa'),
+          },
+        },
+      ],
+      proofs: [proof],
+    }),
+    {
+      authority: w3,
+      capability: capabilities.store.add,
+      principal: Verifier,
+    }
+  )
+
+  assert.equal(r2.error, undefined)
+})
+
+test('invalid proof chain', async () => {
+  const top = await Core.delegate({
+    issuer: alice,
+    audience: bob,
+    capabilities: [
+      {
+        with: alice.did(),
+        can: 'store/add',
+      },
+    ],
+  })
+
+  const proof = await Core.delegate({
+    issuer: bob,
+    audience: mallory,
+    capabilities: [
+      {
+        with: 'did:*',
+        can: '*',
+        nb: {
+          link: '*',
+        },
+      },
+    ],
+    proofs: [top],
+  })
+
+  const result = await access(
+    await Client.delegate({
+      issuer: mallory,
+      audience: w3,
+      capabilities: [
+        {
+          with: alice.did(),
+          can: 'store/add',
+          nb: {
+            link: Link.parse('bafkqaaa'),
+          },
+        },
+      ],
+      proofs: [proof],
+    }),
+    {
+      authority: w3,
+      capability: capabilities.store.add,
+      principal: Verifier,
+    }
+  )
+
+  assert.match(result.toString(), /Expected link to be a CID instead of \*/)
+})
+
+test('restrictions in chain are respected', async () => {
+  const jordan = await ed25519.generate()
+  const top = await Core.delegate({
+    issuer: alice,
+    audience: bob,
+    capabilities: [
+      {
+        with: 'did:*',
+        can: '*',
+      },
+    ],
+  })
+
+  const middle = await Core.delegate({
+    issuer: bob,
+    audience: mallory,
+    capabilities: [
+      {
+        with: 'did:*',
+        can: 'dev/*',
+      },
+    ],
+    proofs: [top],
+  })
+
+  const proof = await Core.delegate({
+    issuer: mallory,
+    audience: jordan,
+    capabilities: [
+      {
+        with: 'did:*',
+        can: '*',
+      },
+    ],
+    proofs: [middle],
+  })
+
+  const boom = await access(
+    await Client.delegate({
+      issuer: jordan,
+      audience: w3,
+      capabilities: [
+        {
+          with: alice.did(),
+          can: 'store/add',
+          nb: {
+            link: Link.parse('bafkqaaa'),
+          },
+        },
+      ],
+      proofs: [proof],
+    }),
+    {
+      authority: w3,
+      // @ts-expect-error - tries to unify incompatible capabilities
+      capability: capabilities.store.add.or(capabilities.dev.ping),
+      principal: Verifier,
+    }
+  )
+
+  assert.match(
+    boom.toString(),
+    /Unauthorized/,
+    'should only allow dev/* capabilities'
+  )
+
+  const ping = capabilities.dev.ping.invoke({
+    issuer: jordan,
+    audience: w3,
+    with: alice.did(),
+    nb: {
+      message: 'hello',
+    },
+    proofs: [proof],
+  })
+
+  const result = await access(await ping.delegate(), {
```

Comment:
see https://github.com/web3-storage/ucanto/pull/218/files#r1100550414

IMHO it's not obvious to me from ucan-wg/spec that `middle` `dev/*` should authorize this, because spec doesn't say anything about `dev/*` syntax (nor `did:*`)

Reply:
So delegation chain we have here is as follows (where top on is ambient authority by ownership, which is also something UCAN spec has no opinion on)

```
{ with: did:key:zAli, can: * }
         ^
         :   - by Ownership to Alice
{ with: did:*, can: * }            =>  { with: did:key:zAli, can: * }
         ^
         |   - by Alice to Bob
{ with: did:*, can: dev/* }        => { with: did:key:zAli, can: dev/* }
         ^
         |   - by Bob to Mallory
{ with: did:*, can: * }            => { with: did:key:zAli, can: dev/* }
         ^
         |   - by Mallory to W3
{ with: did:key:zAli, can: 'dev/ping', nb: { message: 'hello'}  }
```

If you follow the chain you can see all the constraints that had been imposed at every layer (spec calls it diminishing) yet despite all the the constraints bottom is still true.

Now one could argue that Bob is escalating when delegating to Mallory, yet I think it goes against intuition and there is a discussion to spec this pattern because otherwise we have keep recreating chains every time some agent gains more capabilities, for details see:
https://github.com/ucan-wg/spec/issues/130
https://github.com/ucan-wg/spec/issues/131